xbmc/utils/ScreenSaverUtils.cpp

```diff
@@ -0,0 +1,390 @@
+  }
+  else
+  {
+    setting->SetDefault("screensaver.xbmc.builtin.dim");
+  }
+}
+
+CScreenSaverUtils::~CScreenSaverUtils()
+{
+  StopThread();
+
+  CServiceBroker::GetSettingsComponent()->GetSettings()->UnregisterCallback(this);
+
+  Deactivate();
+
+  m_osScreenSaverInhibitor.Release();
+}
+
+void CScreenSaverUtils::Process()
+{
+  while (!m_bStop)
+  {
+    auto winSystem = CServiceBroker::GetWinSystem();
+
+    if (g_application.GetAppPlayer().IsPlayingVideo() &&
+        !g_application.GetAppPlayer().IsPaused())
+    {
+      ResetTimer();
+    }
+    else if (g_application.GetAppPlayer().IsPlayingAudio() &&
+             CServiceBroker::GetGUI()->GetWindowManager().GetActiveWindow() == WINDOW_VISUALISATION &&
+             !CServiceBroker::GetSettingsComponent()->GetSettings()->GetString(CSettings::SETTING_MUSICPLAYER_VISUALISATION).empty())
+    {
+      ResetTimer();
+    }
+    else if (!m_screenSaverActive || (winSystem && !winSystem->GetDPMSManager()->IsActive()))
+    {
+      auto settingsComponent = CServiceBroker::GetSettingsComponent();
+      if (settingsComponent)
+      {
+        auto settings = settingsComponent->GetSettings();
+        if (settings)
+        {
+          int screenSaverSeconds = settings->GetInt(CSettings::SETTING_SCREENSAVER_TIME) * 60;
+          int dpmsSeconds = settings->GetInt(CSettings::SETTING_POWERMANAGEMENT_DISPLAYSOFF) * 60;
+
+          int elapsed = m_screenSaverTimer.IsRunning() ? m_screenSaverTimer.GetElapsedSeconds() : 0;
+
+          if (elapsed > screenSaverSeconds && !m_screenSaverActive)
+          {
+            Activate(false);
+          }
+
+          if (elapsed > dpmsSeconds && !winSystem->GetDPMSManager()->IsActive())
+          {
+            winSystem->GetDPMSManager()->Activate();
+          }
+        }
+      }
+    }
+
+    Sleep(1000);
+  }
+}
+
+void CScreenSaverUtils::ResetTimer()
+{
+  m_screenSaverTimer.StartZero();
+}
+
+void CScreenSaverUtils::StopTimer()
+{
+  m_screenSaverTimer.Stop();
+}
+
+void CScreenSaverUtils::Deactivate()
+{
+  auto winSystem = CServiceBroker::GetWinSystem();
+  if (!winSystem)
+    return;
+
+  winSystem->GetDPMSManager()->Deactivate();
+
+  if (!m_screenSaverActive)
+    return;
+
+  if (m_timeSinceActivation.GetElapsedSeconds() <= 1)
+    return;
+
+  m_timeSinceActivation.Stop();
+  ResetTimer();
+  m_screenSaverActive = false;
+  CServiceBroker::GetAnnouncementManager()->Announce(ANNOUNCEMENT::GUI, "xbmc", "OnScreensaverDeactivated");
+
+  auto settingsComponent = CServiceBroker::GetSettingsComponent();
+
+  if (!settingsComponent)
+    return;
+
+  auto settings = settingsComponent->GetSettings();
+
+  if (!settings)
+    return;
+
+  std::string screensaver = settings->GetString(CSettings::SETTING_SCREENSAVER_MODE);
+
+  if (screensaver == "screensaver.xbmc.builtin.dim" ||
+      screensaver == "screensaver.xbmc.builtin.black" ||
+      screensaver.empty())
+  {
+    CServiceBroker::GetGUI()->GetWindowManager().CloseDialogs();
```

Comment:
that's really bad design. GUI methods must only be called by the render thread. I think this is still the case here but far from obvious. One the one hand you use AppMessanger for sending messages from main thread to main thread, on other hand you call from some util class directly into GUI component.

Reply:
Is there another mechanism that can be used? What is the restriction about calling the GUI from the main thread?